lib/embulk/input/jira_api/client.rb

```diff
@@ -7,31 +7,83 @@ module Embulk
   module Input
     module JiraApi
       class Client
-        PARALLEL_THREAD_COUNT = 50
-        SEARCH_TIMEOUT_SECONDS = 5
-        SEARCH_ISSUES_TIMEOUT_SECONDS = 60
+        MAX_CONCURRENT_REQUESTS = 50
+        # Normal http request timeout is 300s
+        SEARCH_ISSUES_TIMEOUT_SECONDS = 300
         DEFAULT_SEARCH_RETRY_TIMES = 10
 
         def self.setup(&block)
           Jiralicious.configure(&block)
           new
         end
 
+        def find_at_key(array, key)
+          
+        end
+
         def search_issues(jql, options={})
-          timeout_and_retry(SEARCH_ISSUES_TIMEOUT_SECONDS) do
+          parallel_threads = MAX_CONCURRENT_REQUESTS
+          timeout_and_retry(SEARCH_ISSUES_TIMEOUT_SECONDS * MAX_CONCURRENT_REQUESTS ) do
             issues_raw = search(jql, options).issues_raw
-
             # TODO: below code has race-conditon.
-            Parallel.map(issues_raw, in_threads: PARALLEL_THREAD_COUNT) do |issue_raw|
-              # https://github.com/dorack/jiralicious/blob/v0.4.0/lib/jiralicious/search_result.rb#L32-34
-              issue = Jiralicious::Issue.find(issue_raw["key"])
-              JiraApi::Issue.new(issue)
+            success_items = []
+            fail_items = []
+            errors = []
+            search_issue_count = 0
+            
+            while issues_raw.length > 0 && search_issue_count <= DEFAULT_SEARCH_RETRY_TIMES do
+              search_results = Parallel.map(issues_raw, in_threads: parallel_threads) do |issue_raw|
+                # https://github.com/dorack/jiralicious/blob/v0.4.0/lib/jiralicious/search_result.rb#L32-34
+                begin
+                  issue = Jiralicious::Issue.find(issue_raw["key"])
+                  {
+                    :error => nil,
+                    :result => JiraApi::Issue.new(issue)
+                  }
+                rescue MultiJson::ParseError => e
+                  html = e.message
+                  title = html[%r|<title>(.*?)</title>|, 1] #=> e.g. "Unauthorized (401)"
+                  raise title if title != "Unauthorized (401)"
+                  {
+                    :error => e,
+                    :issue_raw => issue_raw,
+                    :result => nil
+                  }
+                end
+              end
+              search_issue_count += 1
+              for issue_result in search_results do
```

Comment:
In case you want to keep this `for` loop, it can be simplified by `#partition`:
```irb
>> succ, failed = [{ error: 'Some error', obj: 'RawIssue1' }, { obj: 'JiraApi::Issue1' }, { error: 'Some error', obj: 'RawIssue2' }].partition{ |row| row[:error].nil? }
=> [[{:obj=>"JiraApi::Issue1"}], [{:error=>"Some error", :obj=>"RawIssue1"}, {:error=>"Some error", :obj=>"RawIssue2"}]]
>> succ
=> [{:obj=>"JiraApi::Issue1"}]
>> failed
=> [{:error=>"Some error", :obj=>"RawIssue1"}, {:error=>"Some error", :obj=>"RawIssue2"}]
```
Note: let's merge `:result` & `:issue_raw` keys, only`:error` is needed for discrimination.

Reply:
A benefit of merging `:result` & `:issue_raw` is that you can easily `map` (extract inner object), for example:
```irb
>> succ, failed = [{ error: 'Some error', obj: 'RawIssue1' }, { obj: 'JiraApi::Issue1' }, { error: 'Some error', obj: 'RawIssue2' }].partition{ |row| row[:error].nil? }.map{ |arr| arr.map{|r| r[:obj]} }
=> [["JiraApi::Issue1"], ["RawIssue1", "RawIssue2"]]
```

Reply:
I have already try to collect inside Parallel but it got some problems: 
1. We have to changed it to Parallel.each
2. If we dont have a semaphore, we could get in rid of race-conditions
3. Introduce a semaphore could lead to some synchronized action and the code could become a little bit complicated while the current Parallel.map already do it for us.

I also think about using partition and map. However in my code I need : 
1. With success items, I need only the result
2. With fail items, I need both the original data to retry and also the error to escalate to the outside in case we fail
So with partition and map, I could not capture the original error. Hence I would suppose to use the single loop.

Reply:
Per discussed offline, we will go with populating `success` and `failed` list during `Parallel.each`.

Reply:
Done